File: phonetic_matching/matching.py
import sys
import re

discarded = 'AEIHOUWY'

categories = ['AEIOU', 'CGJKQSXYZ', 'BFPVW', 'DT', 'MN', 'H', 'L', 'R']


def categorise(name):
    name = re.sub('[^A-Z]', '', name.upper())
    letters = [name[0]]

    for l in name[1:]:
        if l not in discarded:
            letters.append(l)

    cname = []

    for letter in letters:
        for ci, cat in enumerate(categories):
            if letter in cat:
                cletter = ci
                break

        if len(cname) == 0 or cletter != cname[-1]:
            cname.append(cletter)

    return cname


def match(input_names, sample_list):
    names = []

    cinput_names = []
    for name in input_names:
        cinput_names.append(categorise(name))

    for sample_name in sample_list:
        csample_name = categorise(sample_name)
        if csample_name in cinput_names:
            names.append(sample_name)

    return names


if __name__ == '__main__':
    with sys.stdin as f:
        sample_list = f.read().split('\n')

    for name in match(sys.argv[1:], sample_list):
        print(name)

File: phonetic_matching/matching.js
var fs = require('fs');

var discarded = 'AEIHOUWY'
,   categories = ['AEIOU', 'CGJKQSXYZ', 'BFPVW', 'DT', 'MN', 'H', 'L', 'R'];


function categorise (name) {
    var letters, cname, cletter, li, ci, i;

    name = name.toUpperCase().replace(/[^A-Z]/g, '');
    letters = [name[0]];

    for (i = 1; i < name.length; i++) {
        if (discarded.indexOf(name[i]) == -1) {
            letters.push(name[i]);
        }
    }

    cname = [];
    for (li in letters) {
        for (ci in categories) {
            if (categories[ci].indexOf(letters[li]) > -1) {
                cletter = ci;
                break;
            }
        }

        if (cname.length === 0 || cletter != cname.slice(-1)) {
            cname.push(cletter);
        }
    }

    return cname;
}


function match (input_names, sample_list) {
    var names, csample_name, cinput_names, i;

    cinput_names = [];
    for (i in input_names) {
        cinput_names.push(categorise(input_names[i]));
    }

    names = [];
    for (i in sample_list) {
        csample_name = categorise(sample_list[i]);
        if (cinput_names.join().indexOf(csample_name) > -1) {
            names.push(sample_list[i]);
        }
    }

    return names;
}


function main () {
    var data = '';
    process.stdin.on('readable', function () {
        data += process.stdin.read();
    })
    .on('end', function () {
        data = data.match(/[^\r\n]+/g);
        console.log(match(process.argv.slice(2), data).join('\n'));
    });
}


if (require.main === module) main();

exports.match = match;

